---
title: 'xenograft snRNA-seq integration and normalization'
output:
html_document:
theme: united
pdf_document: default
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---
  
  ```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  fig.width = 10,
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

```{r data, message=FALSE}
library(Seurat)
options(Seurat.object.assay.version = "v5")
library(patchwork)
library(Matrix)
library(tidyverse)
library(plyr)
library(ggpubr)
library(HGNChelper)
library(EnhancedVolcano)
library(ggbreak) 
theme_set(theme_bw())
options(future.globals.maxSize = 1e9)
options(Seurat.object.assay.version = "v5")

```

```{r define.paths, message=FALSE, warning=FALSE}
# working directory
working_dir="/Users/zacc/USyd/ASAP/snrna_seq/analysis"
setwd(working_dir)

# path of cellranger output
cellranger_out="/Users/zacc/USyd/ASAP/snrna_seq/cellranger_out"
```


```{r cellranger.data, results='hide', message=FALSE, fig.keep='none'}
# Read in multiple cellranger outputs

# get list of datasets
matrix_dir = "outs/filtered_feature_bc_matrix"
cr_dirs <- list.dirs(cellranger_out)
cr_dirs <- as.list(cr_dirs[grep(matrix_dir,cr_dirs)])

# get 12w and 6w 
cr_dirs <- cr_dirs[c(1,4)]

# read in experiments
cr_dfs <- lapply(cr_dirs, function(x) Read10X(unlist(x)))
names(cr_dfs) <- gsub("/","",gsub(matrix_dir,"",gsub(cellranger_out,"",cr_dirs)))

# create seurat objects - note I couldn't work out how to unlist & subset on a list of s4 objects in lapply, hence loop.
cr_so <- list()
for (i in 1:length(cr_dfs)) {
  cr_so[[i]] <- CreateSeuratObject(counts = unlist(cr_dfs[i])[[1]], project = names(cr_dfs[i]))
}

# remove genes with no counts
cr_so <- lapply(X = cr_so, FUN = function(x) {
  x = x[,unname(which(colSums(GetAssayData(x))!=0))]
})

```


```{r combine.lists, results='hide', message=FALSE, fig.keep='none'}
# combine seurat lists
seurat_list <- c(cr_so)
# get/ set names
seurat_names <- gsub(".*2_","",c(names(cr_dfs)))
# get/ set names
seurat_names_2 <- gsub("hm","Mouse",gsub("hr","Rat",gsub(".*2_","",c(names(cr_dfs)))))
```

```{r qc.sample.filter, results='hide', message=FALSE, fig.keep='none'}
# Percentage of reads that map to the mitochondrial genome
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
    x <- AddMetaData(object = x, metadata = PercentageFeatureSet(x, pattern = "-MT-|-Mt-|-mt-"), col.name = "percent.mt")

})

# Visualize the distribution of mitochondrial gene expression detected per cell
plot.list <- list()
for (i in 1:2) {
  plot.list[[i]] <- seurat_list[[i]]@meta.data %>% 
    ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + xlab("mtRNA (%)") + xlim(0,0.15) +
    theme_classic()  + theme(legend.position = "none") +
    geom_vline(xintercept = 0.1, lty =2)
  
}
arrange <- ggarrange(plotlist=plot.list, nrow=3, ncol=4)
ggsave("cellranger_mtprc_dist.png", arrange)


# Visualize the distribution of nFeature_RNA per cell
plot.list <- list()
for (i in 1:2) {
  plot.list[[i]] <- seurat_list[[i]]@meta.data %>% 
    ggplot(aes(color=orig.ident, x=nFeature_RNA, fill=orig.ident)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic()  + theme(legend.position = "none") +
    geom_vline(xintercept = 2000, lty =2)
  
}
arrange <- ggarrange(plotlist=plot.list, nrow=3, ncol=4)
ggsave("cellranger_nFeature_dist.png", arrange)


#Filter cells that have unique filter counts < 2000 & >0.1% mitochondrial counts
seurat_list <- lapply(X = seurat_list, FUN = function(x) {
    x <- subset(x, subset = nFeature_RNA >= 2000 & percent.mt < 0.1)

})

# mean mtRNA percentage and nCount after filtering
lapply(seurat_list,function(x) mean(x$nFeature_RNA))
lapply(seurat_list,function(x) sd(x$nFeature_RNA))
lapply(seurat_list,function(x) mean(x$percent.mt))
```


```{r qc.gene.filter, results='hide', message=FALSE, fig.keep='none'}
# Visualize the number UMIs/transcripts per cell 
plot.list <- list()
for (i in 1:length(seurat_list)) {
  plot.list[[i]] <- seurat_list[[i]]@meta.data %>% 
    ggplot(aes(x=nCount_RNA)) + 
    geom_density(alpha = 0.2) + 
    scale_x_log10() + 
    theme_classic() +
    ylab("Cell density") +
    geom_vline(xintercept = 500)
}
ggarrange(plotlist=plot.list, nrow=2, ncol=4, labels = seurat_names)

plot.list <- list()
for (i in 1:length(seurat_list[1:2])) {
  plot.list[[i]] <- seurat_list[[i]]@meta.data %>% 
    ggplot(aes(x=nFeature_RNA)) + 
    geom_density(alpha = 0.2) + 
    theme_classic() +
    scale_x_log10() + 
    geom_vline(xintercept = 300)
}
ggarrange(plotlist=plot.list, nrow=1, ncol=4, labels = seurat_names_2)


```

```{r integrate, results='hide', message=FALSE, fig.keep='none'}
# from Seurat 5.01 vignette Compiled: November 16, 2023
# merge non-normalized data
seurat_merge <- merge(seurat_list[[1]], y = seurat_list[[2]], add.cell.ids = c("12w", "6w"))

# join layers after integration
seurat_merge[["RNA"]] <- JoinLayers(seurat_merge[["RNA"]])

# split the RNA measurements into layers x samples
seurat_merge[["RNA"]] <- split(seurat_merge[["RNA"]], f = seurat_merge$orig.ident)

# run standard anlaysis workflow without integration
seurat_merge <- NormalizeData(seurat_merge)
seurat_merge <- FindVariableFeatures(seurat_merge)
seurat_merge <- ScaleData(seurat_merge)
seurat_merge <- RunPCA(seurat_merge)
seurat_merge <- FindNeighbors(seurat_merge, dims = 1:30, reduction = "pca")
seurat_merge <- FindClusters(seurat_merge, resolution = 2, cluster.name = "unintegrated_clusters")
seurat_merge <- RunUMAP(seurat_merge, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")

# Visualization
DimPlot(seurat_merge, reduction = "umap.unintegrated", group.by = c("orig.ident", "seurat_clusters"))


# integrate datasets
seurat_merge <- IntegrateLayers(object = seurat_merge, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca",verbose = FALSE)

# re-join layers after integration
seurat_merge[["RNA"]] <- JoinLayers(seurat_merge[["RNA"]])

# re-cluster
seurat_merge <- FindNeighbors(seurat_merge, reduction = "integrated.cca", dims = 1:30)
seurat_merge <- FindClusters(seurat_merge, resolution = 1)
seurat_merge <- RunUMAP(seurat_merge, dims = 1:30, reduction = "integrated.cca")

# Visualization
DimPlot(seurat_merge, reduction = "umap", group.by = c("orig.ident", "seurat_clusters"))

```

```{r normalise, results='hide', message=FALSE, fig.keep='none'}
seurat_merge.sct <- SCTransform(seurat_merge)

```

```{r bin.human.rodent, results='hide', message=FALSE, fig.keep='none'}
# function to calculate the proportions of counts mapping to human and rodent genes.
quant_human_rodent <- function(res_mat) {
  rodent_inds <- !grepl("GRCh38",row.names(res_mat[["SCT"]]$counts))
  human_inds <- grepl("GRCh38",row.names(res_mat[["SCT"]]$counts))
  cell_species <- tibble(n_rodent_umi = Matrix::colSums(res_mat[["SCT"]]$counts[rodent_inds,]),
                         n_human_umi = Matrix::colSums(res_mat[["SCT"]]$counts[human_inds,]),
                         tot_umi = Matrix::colSums(res_mat[["SCT"]]$counts),
                         prop_rodent = n_rodent_umi / tot_umi,
                         prop_human = n_human_umi / tot_umi)
  
}

# run and add to metadata
seurat_merge.sct <- AddMetaData(seurat_merge.sct, metadata = quant_human_rodent(seurat_merge.sct))

# plot of human proportions distribution
plot.list <- list()
orig_ident <- unique(seurat_merge.sct$orig.ident)
for (i in 1:length(orig_ident)) {
  plot.list[[i]] <- seurat_merge.sct@meta.data[seurat_merge.sct$orig.ident == orig_ident[i],] %>% 
    ggplot(aes(x=prop_human)) + 
    geom_density(alpha = 0.2) + 
    #scale_y_break(c(100,800)) +
    theme_classic() +
    xlab("Human Proportion") + xlim(-0.05,1.05) +
    geom_vline(xintercept = 0.9, lty =2) 
}

arrange <- ggarrange(plotlist=plot.list, nrow=2, ncol=4, widths = c(2,2))
ggsave("humanprop_dist.png", arrange)

# bin into human or rodent using 0.9 human prop threshold
seurat_merge.sct$origin <- seurat_merge.sct@meta.data$prop_human > 0.9
seurat_merge.sct$origin[seurat_merge.sct$origin == "TRUE"] <- "human"
seurat_merge.sct$origin[seurat_merge.sct@meta.data$prop_human < 0.01] <- "rodent"
seurat_merge.sct$origin[seurat_merge.sct$origin == "FALSE"] <- "unknown"

# Visualization
p1 <- DimPlot(seurat_merge.sct, reduction = "umap", group.by = c("orig.ident"))
p2 <- DimPlot(seurat_merge.sct, reduction = "umap", group.by = c("seurat_clusters"))
p3 <- DimPlot(seurat_merge.sct, reduction = "umap", group.by = c("origin"))

ggsave("snrnaseq_lowlvl_cluster_orig.ident.pdf",p1)
ggsave("snrnaseq_lowlvl_cluster_seurat_clusters.pdf",p2)
ggsave("snrnaseq_lowlvl_cluster_origin.pdf",p3)

# calculate proportions of each sample with human nuclei
table(seurat_merge.sct$origin,seurat_merge.sct$orig.ident)[1,]/
table(seurat_merge.sct$orig.ident) * 100

```

```{r select.human, results='hide', message=FALSE}

# get human genes
genes.use <- row.names(seurat_merge.sct)[grep("GRCh38",row.names(seurat_merge.sct))]

# get human cells
ident.use <- row.names(seurat_merge.sct@meta.data)[seurat_merge.sct@meta.data$origin == "human"]

# create new seurat object
subset.matrix <- seurat_merge.sct@assays$RNA$data[genes.use, ident.use] # Pull the raw expression matrix from the original Seurat object containing only the genes of 
# strip human identifiers
row.names(subset.matrix) <- gsub("GRCh38--","",row.names(subset.matrix))

# subset samples of interest
seurat_human <- CreateSeuratObject(subset.matrix) # Create a new Seurat object with just the genes of interest
orig.ident <- seurat_merge.sct@meta.data[ident.use,] # Pull the identities from the original Seurat object as a data.frame
seurat_human <- AddMetaData(object = seurat_human, metadata = orig.ident) # Add the idents to the meta.data slot
Idents(object = seurat_human) <- ident.use # Assign identities for the new Seurat object

# split the RNA measurements into layers x samples
seurat_human[["RNA"]] <- split(seurat_human[["RNA"]], f = seurat_human$orig.ident)

# run standard anlaysis workflow without integration
seurat_human <- NormalizeData(seurat_human)
seurat_human <- FindVariableFeatures(seurat_human)
seurat_human <- ScaleData(seurat_human)
seurat_human <- RunPCA(seurat_human)
seurat_human <- FindNeighbors(seurat_human, dims = 1:30, reduction = "pca")
seurat_human <- FindClusters(seurat_human, resolution = 2, cluster.name = "unintegrated_clusters")
seurat_human <- RunUMAP(seurat_human, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")

# integrate datasets
seurat_human <- IntegrateLayers(object = seurat_human, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca",verbose = FALSE, k.weight = 50)

# re-join layers after integration
seurat_human[["RNA"]] <- JoinLayers(seurat_human[["RNA"]])

# re-cluster
seurat_human <- FindNeighbors(seurat_human, reduction = "integrated.cca", dims = 1:30)
seurat_human <- FindClusters(seurat_human, resolution = 1)
seurat_human <- RunUMAP(seurat_human, dims = 1:30, reduction = "integrated.cca")

# Visualization
p1 <- DimPlot(seurat_human, reduction = "umap", group.by = c("orig.ident", "seurat_clusters"))
ggsave("snrnaseq_lowlvl_cluster_humancells.pdf",p1)


# plot metrics for each cluster
metrics <-  c("nFeature_RNA", "nCount_RNA", "percent.mt")

p4 <- FeaturePlot(seurat_human, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4)

ggsave("umap_seqmetrics_human.png", p4)

# Extract identity and sample information from seurat object to determine the number of cells per cluster per sample
n_cells <- FetchData(seurat_human, 
                     vars = c("ident", "orig.ident")) %>%
        dplyr::count(ident, orig.ident) %>%
        tidyr::spread(ident, n)
n_cells


# format metadata
seurat_human@meta.data$week <- seurat_human@meta.data$orig.ident
seurat_human@meta.data$week[seurat_human@meta.data$orig.ident == "run_HJ377DRX2_12w_hr" ] <- 12
seurat_human@meta.data$week[seurat_human@meta.data$orig.ident == "run_HLNGJDRX2_6w_hr" ] <- 6
seurat_human$week <- factor(x = object@ident, levels = c(6,12))

# save seurat object
#  saveRDS(seurat_human, file = "xeno_human_snrnaseq_seurat_070524.rds")

```


```{r cell.annotation.CHETAH, results='hide', message=FALSE}
library(CHETAH)

## Make SingleCellExperiments 
load("/Volumes/research-data/PRJ-ASAPbioin/Bioinformics/datasets/seurat_objects/kamath_seurat_subcells.Rdata")
kamath.sce <- as.SingleCellExperiment(seurat_obj_subcells)
colData(kamath.sce)$celltypes <- colData(kamath.sce)$Cell_Type

seurat_human <- readRDS("/Users/zacc/USyd/ASAP/snrna_seq/analysis/xeno_human_snrnaseq_seurat_070524.rds")
input_counts <- seurat_human@assays$RNA$counts

input <- SingleCellExperiment(assays = list(counts = input_counts),
                              reducedDims = SimpleList(UMAP = seurat_human[["umap"]]@cell.embeddings))

## Run CHETAH
input <- CHETAHclassifier(input = input, ref_cells = kamath.sce)

## Plot the classification

p1 <- PlotCHETAH(input, pt.size = 4,return =TRUE)
ggsave("chetah_cellclass.png", p1,width = 25, height = 10)

p2 <- PlotCHETAH(input = input, interm = TRUE, pt.size = 4,return =TRUE)
ggsave("chetah_cellclass_interm.png", p2,width = 25, height = 10)

## Extract celltypes:
celltypes <- input$celltype_CHETAH

meta_dat <- seurat_human@meta.data
names(celltypes) == row.names(meta_dat)
meta_dat$celltypes <- celltypes

table(meta_dat$seurat_clusters,meta_dat$celltypes)

```


```{r cell.annotation, results='hide', message=FALSE}
seurat_human <- readRDS("/Users/zacc/USyd/ASAP/snrna_seq/analysis/xeno_human_snrnaseq_seurat_070524.rds")

# set assy to RNA for cell annotation
DefaultAssay(merge.combined.sct) <- "RNA"
seurat_human <- ScaleData(seurat_human, verbose = FALSE)


# load gene set preparation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R")
# load cell type annotation function
source("https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R")

# DB file
db_ = "https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_full.xlsx";
tissue = "Brain"
# e.g. Immune system,Pancreas,Liver,Eye,Kidney,Brain,Lung,Adrenal,Heart,Intestine,Muscle,Placenta,Spleen,Stomach,Thymus 

# # prepare gene sets
gs_list = gene_sets_prepare(db_, tissue)


# get cell-type by cell matrix
# es.max = sctype_score(scRNAseqData = merge.combined.sct[["RNA"]]@scale.data, scaled = TRUE,
#                        gs = gs_list$gs_positive)
# NOTE:  the sctype_score function is only working when run outside function command??? I dont know why??
# Hence running outside function command
scRNAseqData = seurat_human@assays$RNA$scale.data
scaled = TRUE
gs = gs_list$gs_positive
gs2 = NULL
gene_names_to_uppercase = 0

   # check input matrix
    if(!is.matrix(scRNAseqData)){
        warning("scRNAseqData doesn't seem to be a matrix")
    } else {
        if(sum(dim(scRNAseqData))==0){
            warning("The dimension of input scRNAseqData matrix equals to 0, is it an empty matrix?")
        }
    }
    
    # marker sensitivity
    marker_stat = sort(table(unlist(gs)), decreasing = T); 
    marker_sensitivity = data.frame(score_marker_sensitivity = scales::rescale(as.numeric(marker_stat), to = c(0,1), from = c(length(gs),1)),
                                    gene_ = names(marker_stat), stringsAsFactors = !1)

    # convert gene names to Uppercase
    if(gene_names_to_uppercase){
        rownames(scRNAseqData) = toupper(rownames(scRNAseqData));
    }
    
    # subselect genes only found in data
    names_gs_cp = names(gs); names_gs_2_cp = names(gs2);
    gs = lapply(1:length(gs), function(d_){ 
        GeneIndToKeep = rownames(scRNAseqData) %in% as.character(gs[[d_]]); rownames(scRNAseqData)[GeneIndToKeep]})
    gs2 = lapply(1:length(gs2), function(d_){ 
        GeneIndToKeep = rownames(scRNAseqData) %in% as.character(gs2[[d_]]); rownames(scRNAseqData)[GeneIndToKeep]})
    names(gs) = names_gs_cp; names(gs2) = names_gs_2_cp;
    cell_markers_genes_score = marker_sensitivity[marker_sensitivity$gene_ %in% unique(unlist(gs)),]

    # z-scale if not
    if(!scaled) Z <- t(scale(t(scRNAseqData))) else Z <- scRNAseqData
    
    # multiple by marker sensitivity
    for(jj in 1:nrow(cell_markers_genes_score)){
        Z[cell_markers_genes_score[jj,"gene_"], ] = Z[cell_markers_genes_score[jj,"gene_"], ] * cell_markers_genes_score[jj, "score_marker_sensitivity"]
    }
    
    # subselect only with marker genes
    Z = Z[unique(c(unlist(gs),unlist(gs2))), ]
    
    # combine scores
    es = do.call("rbind", lapply(names(gs), function(gss_){ 
        sapply(1:ncol(Z), function(j) {
            gs_z = Z[gs[[gss_]], j]; gz_2 = Z[gs2[[gss_]], j] * -1
            sum_t1 = (sum(gs_z) / sqrt(length(gs_z))); sum_t2 = sum(gz_2) / sqrt(length(gz_2));
            if(is.na(sum_t2)){
                sum_t2 = 0;
            }
            sum_t1 + sum_t2
        })
    })) 
    
    dimnames(es) = list(names(gs), colnames(Z))
    es.max <- es[!apply(is.na(es) | es == "", 1, all),] # remove na rows


# merge by cluster
cL_resutls = do.call("rbind", lapply(unique(seurat_human@meta.data$seurat_clusters), function(cl){
    es.max.cl = sort(rowSums(es.max[ ,rownames(seurat_human@meta.data[seurat_human@meta.data$seurat_clusters==cl, ])]), decreasing = !0)
    head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(seurat_human@meta.data$seurat_clusters==cl)), 10)
}))

sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  

# define thresholds
thresholds <- sctype_scores$ncells/4

# plot 
dplot <- cL_resutls[cL_resutls$cluster == 0,]
p1 <- ggplot(data = dplot, aes(x = reorder(type, scores), y = scores)) +
  geom_bar(stat = "identity") + xlab("Cell type") +
    theme_classic() +
    geom_hline(yintercept = thresholds[1], lty =2) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) + ggtitle("Cluster 0")

dplot <- cL_resutls[cL_resutls$cluster == 1,]
p2 <- ggplot(data = dplot, aes(x = reorder(type, scores), y = scores)) +
  geom_bar(stat = "identity") + xlab("Cell type") +
    theme_classic() +
    geom_hline(yintercept = thresholds[2], lty =2) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) + ggtitle("Cluster 1")

dplot <- cL_resutls[cL_resutls$cluster == 2,]
p3 <- ggplot(data = dplot, aes(x = reorder(type, scores), y = scores)) +
  geom_bar(stat = "identity") + xlab("Cell type") +
    theme_classic() +
    geom_hline(yintercept = thresholds[2], lty =2) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) + ggtitle("Cluster 2")

arrange <- ggarrange(plotlist=list(p1,p2,p3), nrow=2, ncol=2, widths = c(2,2))
ggsave("sctype_cl0.1.2.png", arrange)


# set low-confident (low ScType score) clusters to "unknown"
sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < thresholds] = "Unknown"
print(sctype_scores[,1:3])


```


```{r assign.cluster.cell = 'hide', message=FALSE}
# Assign clusters name based on CHETAH and scType
seurat_human@meta.data$cell_type <- as.character(seurat_human@meta.data$seurat_clusters)
seurat_human@meta.data$cell_type[seurat_human@meta.data$cell_type != "2"] <- "non-DA"
seurat_human@meta.data$cell_type[seurat_human@meta.data$cell_type == "2"] <- "DA"

# save seurat object
#  saveRDS(seurat_human, file = "xeno_human_snrnaseq_seurat_070524.rds")
```


```{r DEG.voom = 'hide', message=FALSE}
seurat_human <- readRDS("/Users/zacc/USyd/ASAP/snrna_seq/analysis/xeno_human_snrnaseq_seurat_140324.rds")
DefaultAssay(seurat_human) <- "RNA"
library(edgeR)

# create list to collect results 
res <- list()


## 1) DEGs; Xenograft DA v non-DA
# extract expression values and meta data
exp_dat <- as.data.frame(seurat_human@assays$RNA$data) # note these are the raw counts
targ <- seurat_human@meta.data

# subset for var of interest
dge <- DGEList(exp_dat, group = targ$cell_type)
dge <- calcNormFactors(dge)
design <- model.matrix( ~ cell_type, data=targ)
vm <- voom(dge, design = design, plot = TRUE)
fit <- lmFit(vm, design = design)
fit <- eBayes(fit)
tt <- topTable(fit, n = Inf, adjust.method = "BH")

tt$contrast <- "DA.nonDA"
res[[1]] <- tt

which(row.names(tt) %in% c("NTN1","SEMA3A","WNT5A"))

## 2) DEGs; Xenograft DA (KCNJ6+/TH+) v non-DA 
# extract expression values and meta data
exp_dat <- as.data.frame(seurat_human@assays$RNA$data) # note these are the raw counts
targ <- seurat_human@meta.data
targ$cell_type_exp <- c(targ$cell_type == "DA" & exp_dat["KCNJ6",] > 0 & exp_dat["TH",] > 0)
targ$cell_type_exp[targ$cell_type_exp == "TRUE"] <- "DA.kcnj6.th"
targ$cell_type_exp[targ$cell_type == "non-DA"] <- "non-DA"

exp_dat <- exp_dat[,targ$cell_type_exp != "FALSE"]
targ <- targ[targ$cell_type_exp != "FALSE",]

# subset for var of interest
dge <- DGEList(exp_dat, group = targ$cell_type_exp)
dge <- calcNormFactors(dge)
design <- model.matrix( ~ cell_type_exp, data=targ)
vm <- voom(dge, design = design, plot = TRUE)
fit <- lmFit(vm, design = design)
fit <- eBayes(fit)
tt <- topTable(fit, n = Inf, adjust.method = "BH")

tt$contrast <- "DA_kcnj6_th.nonDA"
res[[2]] <- tt

which(row.names(tt) %in% c("NTN1","SEMA3A","WNT5A","DCC"))
tt[c("NTN1","SEMA3A","WNT5A","DCC"),]


## 3) DEGs; cluster x time
# extract expression values and meta data
exp_dat <- as.data.frame(seurat_human@assays$RNA$data) # note these are the raw counts
targ <- seurat_human@meta.data

# subset for var of interest
dge <- DGEList(exp_dat, group = targ$cell_type)
dge <- calcNormFactors(dge)
design <- model.matrix( ~ 0 + cell_type * week, data=targ)
colnames(design) <- make.names(colnames(design))
vm <- voom(dge, design = design, plot = TRUE)
fit <- lmFit(vm, design = design)
cont.matrix <- makeContrasts(A="(cell_typeDA - week12) - (cell_typenon.DA - week12)",levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
vfit2 <- eBayes(fit2)
tt <- topTable(vfit2, n = Inf, adjust.method = "BH")

tt$contrast <- "DA.nonDA_time"
res[[3]] <- tt

which(row.names(tt) %in% c("NTN1","SEMA3A","WNT5A","DCC"))
tt[c("NTN1","SEMA3A","WNT5A","DCC"),]


## 4) DEGs; cluster (KCNJ6+/TH+) x time
# extract expression values and meta data
exp_dat <- as.data.frame(seurat_human@assays$RNA$data) # note these are the raw counts
targ <- seurat_human@meta.data
targ$cell_type_exp <- c(targ$cell_type == "DA" & exp_dat["KCNJ6",] > 0 & exp_dat["TH",] > 0)
targ$cell_type_exp[targ$cell_type_exp == "TRUE"] <- "DA.kcnj6.th"
targ$cell_type_exp[targ$cell_type == "non-DA"] <- "non-DA"

exp_dat <- exp_dat[,targ$cell_type_exp != "FALSE"]
targ <- targ[targ$cell_type_exp != "FALSE",]

# subset for var of interest
dge <- DGEList(exp_dat, group = targ$cell_type_exp)
dge <- calcNormFactors(dge)
design <- model.matrix( ~ 0 + cell_type_exp * week, data=targ)
colnames(design) <- make.names(colnames(design))
vm <- voom(dge, design = design, plot = TRUE)
fit <- lmFit(vm, design = design)
cont.matrix <- makeContrasts(A="(cell_type_expDA.kcnj6.th - week12) - (cell_type_expnon.DA - week12)",levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
vfit2 <- eBayes(fit2)
tt <- topTable(vfit2, n = Inf, adjust.method = "BH")

tt$contrast <- "DA_kcnj6_th.nonDA_time"
res[[4]] <- tt

which(row.names(tt) %in% c("NTN1","SEMA3A","WNT5A","DCC"))
tt[c("NTN1","SEMA3A","WNT5A","DCC"),]
t1 <- tt[tt$logFC > 0,]
which(row.names(t1) %in% c("NTN1","SEMA3A","WNT5A","DCC"))



# write to excel
library(openxlsx)
tmp <- as.data.frame(res[[1]])
tmp$Gene <- row.names(tmp)
write.xlsx(tmp, file = "deg_voom_DA.nonDA.xlsx")

tmp <- as.data.frame(res[[2]])
tmp$Gene <- row.names(tmp)
write.xlsx(tmp, file = "deg_voom_DA_kcnj6_th.nonDA.xlsx")

tmp <- as.data.frame(res[[3]])
tmp$Gene <- row.names(tmp)
write.xlsx(tmp, file = "deg_voom_cluster.time.xlsx")

tmp <- as.data.frame(res[[4]])
tmp$Gene <- row.names(tmp)
write.xlsx(tmp, file = "deg_voom_cluster_kcnj6_th.time.xlsx", row.names = TRUE)









####
## Combined p-values of cluster 0-1 DEGs and clusters 6-12 transition DEGs
# Function to combine p-values using Stouffer's Method
combine_pvalues_stouffer <- function(p_values) {
  # Convert p-values to Z-scores
  z_scores <- qnorm(p_values / 2, lower.tail = FALSE)
  
  # Calculate combined Z-score
  combined_z <- sum(z_scores) / sqrt(length(z_scores))
  
  # Convert combined Z-score back to a p-value
  combined_p <- 2 * pnorm(-abs(combined_z))
  
  return(combined_p)
}

x <- as.data.frame(res[[1]])
y <- as.data.frame(res[[2]])

x$Gene <- row.names(x)
y$Gene <- row.names(y)
tmp1 <- merge(x,y, by="Gene")

#tmp1 <- tmp1[tmp1$logFC.x > 0, ] # want to look at +logFC i.e. greater in cluster 1
#tmp1 <- tmp1[tmp1$logFC.y < 0, ] # want to look at -logFC i.e. increased in cluster 1, 6 to 12 weeks but not cluster 0

# calc stouffer
for (z in 1:nrow(tmp1)){
      tmp1$stouffer_p[z] <- combine_pvalues_stouffer(c(tmp1$adj.P.Val.x[z],tmp1$adj.P.Val.y[z]))
  }

tmp1 <- tmp1[order(tmp1$stouffer_p),]
which(tmp1$Gene %in% "NTN1")
merged_res <- tmp1

# ## 2) DEGs; snRNAseq
# # extract expression values and meta data
# d2 <- as.data.frame(seurat_obj_subcells@assays$RNA$counts)
# m2 <- seurat_obj_subcells@meta.data
# exp_dat <- d2[,m2$Cell_Type %in% c("CALB1_CALCR","CALB1_CRYM_CCDC68",
#                                    "CALB1_GEM","CALB1_PPP1R17","CALB1_RBP4",
#                                    "CALB1_TRHR","SOX6_AGTR1","SOX6_DDT","SOX6_GFRA2","SOX6_PART1",
#                               unique(m2$Cell_Type)[grep("Ex_|Inh_",unique(m2$Cell_Type))])]
# targ <- m2[colnames(exp_dat),]
# targ$cell_type_broad <- gsub("_.*","",targ$Cell_Type)
# 
# # create design matrix
# options(na.action='na.omit')
# design <- model.matrix(reformulate("~ 0 + cell_type_broad"),
#                        data=targ, 
#                        drop = FALSE)
# # make names
# colnames(design) <- make.names(colnames(design))
# 
# # create expression df & targ with design row.names
# targ <- targ[row.names(design),]
# y <- exp_dat[,row.names(targ)]
# 
# # fit design
# v <- voom(y,design)
# vfit <- lmFit(v)
# 
# # Perform LIMMA contrasts
# cont.matrix <- makeContrasts(B="(cell_type_broadSOX6 + cell_type_broadCALB1) - (cell_type_broadEx + cell_type_broadInh)",
#                              C="(cell_type_broadSOX6) - (cell_type_broadEx + cell_type_broadInh + cell_type_broadCALB1)",levels=design)
# fit2 <- contrasts.fit(vfit, cont.matrix)
# vfit2 <- eBayes(fit2)
# options(digits=3)
# 
# # Select significant DEGs and assign to list
# tmp <- topTable(vfit2,coef = "B",number = Inf,sort.by="P")
# dim(tmp[tmp$adj.P.Val < 0.05,])
# res[[2]] <- tmp
# 
# tmp <- topTable(vfit2,coef = "C",number = Inf,sort.by="P")
# dim(tmp[tmp$adj.P.Val < 0.05,])
# res[[3]] <- tmp

# relabel 
colnames(merged_res) <- gsub(".x",".DAdeg",colnames(merged_res))
colnames(merged_res) <- gsub(".y",".DAdeg6.12",colnames(merged_res))

# plot volcano
dplot <- merged_res
  lab_italics <- paste0("italic('", dplot$Gene, "')")
  selectLab_italics = paste0(
    "italic('",
    c('NTN1','SLC6A3'),
    "')")

  keyvals.colour <- ifelse(
    dplot$logFC.DAdeg > 1, 'red',
      ifelse(dplot$logFC.DAdeg < 1, 'grey',
        'grey1'))
  keyvals.colour[is.na(keyvals.colour)] <- 'grey1'
  names(keyvals.colour)[keyvals.colour == 'red'] <- 'candidate'
  names(keyvals.colour)[keyvals.colour == 'grey'] <- ''

p1 <- EnhancedVolcano(dplot,
    lab = lab_italics,
    x = 'logFC.DAdeg6.12',
    y = 'stouffer_p',
    selectLab = selectLab_italics,
    xlab = bquote(~Log[2]~ 'fold change'),
    colCustom = keyvals.colour,
    pCutoff = 0.05,
    FCcutoff = 1.0,
    pointSize = 3.0,
    labSize = 6.0,
    labCol = 'black',
    labFace = 'bold',
    boxedLabels = TRUE,
    parseLabels = TRUE,
    col = c('black', 'pink', 'purple', 'red3'),
    colAlpha = 4/5,
    legendPosition = 'bottom',
    legendLabSize = 14,
    legendIconSize = 4.0,
    drawConnectors = TRUE,
    widthConnectors = 1.0,
    colConnectors = 'black') + coord_flip()
  
  
pdf("volcano_deg_stouffer.pdf")
p1
dev.off()


## write results to file
names(res) <- c("DEG_Xenograft_KCNJ6.TH_Else_UMI")
#names(res) <- c("DEG_Xenograft_cluster0.cluster1","DEG_SNpc.scRNAseq_DaN.ExcInh","DEG_SNpc.scRNAseq_Sox6.Else") # Note DEG analysis for snRNAseq not working currently, need to review.
#res <- lapply(res, function(x) x[x$adj.P.Val < 0.05,])
res <- lapply(res, function(x) {
  x$Gene <- row.names(x)
  return(x)
  })

## Subset for chiara select
req1 <- read.delim("/Users/zacc/USyd/ASAP/snrna_seq/analysis/chiara_request_100124.txt")
tmp2 <- merge(tmp,req1, by.x = "row.names", by.y = "Gene")
res2 <- list()
res2[[1]] <- tmp2
names(res2) <- c("DEG_Xenograft_cluster1.cluster0")

# write to excel
library(openxlsx)
write.xlsx(res, file = "DEGs_LIMMA_Voom.xlsx", row.names = FALSE)
write.xlsx(res, file = "DEGs_LIMMA_Voom_KCNJ6.TH.xlsx", row.names = FALSE)
write.xlsx(res, file = "DEGs_LIMMA_Voom_KCNJ6.TH_UMI.xlsx", row.names = FALSE)
write.xlsx(res2, file = "DEGs_LIMMA_Voom_request100124.xlsx", row.names = FALSE)


```



```{r plot.marker.genes, results='hide', message=FALSE}
# gene list of interest
my_genes <- c( "KCNJ6", "TH","NTN1","DCC","SEMA3A","WNT5A")

# Select the RNA counts slot to be the default assay
DefaultAssay(seurat_human) <- "RNA"

#Violin plot of markers of interest
p1 <- VlnPlot(seurat_human, features=my_genes, group.by = 'cell_type', split.by = 'week')


ggsave("vln_genes_interest.png", p1)

```





```{r DEG.ontologies = 'hide', message=FALSE}

## 1. Run Gene Ontology
library(enrichR)
#tmp2 <- res[[2]]
tmp2 <- merged_res[merged_res$logFC.x > 0 & merged_res$logFC.y < 0,][1:22,]

# # remove mtRNA for any xenograft to snRNAseq contrasts
# tmp2 <- tmp2[!row.names(tmp2) %in% row.names(tmp2)[grep("MT-",row.names(tmp2))],]
# tmp2 <- tmp2[tmp2$adj.P.Val < 0.05,]

# upreg list
upreg <- row.names(tmp2)[tmp2$logFC < 0 & tmp2$adj.P.Val < 0.05 ]
#upreg <- tmp2$Gene
length(upreg)

# # downreg list
# downreg <- row.names(tmp2)[tmp2$logFC < 0]
# length(downreg)

# select databases
dbs <- c("GO_Molecular_Function_2015", "GO_Cellular_Component_2015", "GO_Biological_Process_2015","Jensen_COMPARTMENTS")

# upreg top 3 passing adj p-val < 0.05
enriched <- enrichr(upreg, dbs)
lapply(enriched, function(x) {
  x <- x[x$Adjusted.P.value < 0.05,]
  x[1:3,]
})

# plot enriched pathways
p1 <- plotEnrich(enriched[["GO_Molecular_Function_2015"]], showTerms = 20, numChar = 40, y = "Count", orderBy = "Adjusted.P.value", title = "GO Molecular Function")
p2 <- plotEnrich(enriched[["GO_Cellular_Component_2015"]], showTerms = 20, numChar = 40, y = "Count", orderBy = "Adjusted.P.value", title = "GO Cellular Component")
p3 <- plotEnrich(enriched[["Jensen_COMPARTMENTS"]], showTerms = 20, numChar = 40, y = "Count", orderBy = "Adjusted.P.value", title = "Jensen COMPARTMENTS")

arrange <- ggarrange(plotlist=list(p1,p2,p3), nrow=2, ncol=2, widths = c(2,2))
# ggsave("GO_enrichment_upreg_DAneuron.png", arrange)

ggsave("GO_Molecular_Function_2015_upreg_DAneuron6.12.png", p1)
ggsave("GO_Cellular_Component_2015_upreg_DAneuron6.12.png", p2)
ggsave("Jensen_COMPARTMENTS_upreg_DAneuron6.12.png", p3)

# write to excel
write.xlsx(enriched[["GO_Molecular_Function_2015"]], file = "DEG_GO_Molecular_Function_2015_upreg_DAneuron6.12.xlsx", row.names = FALSE)
write.xlsx(enriched[["GO_Cellular_Component_2015"]], file = "DEG_GO_Cellular_Component_2015_upreg_DAneuron6.12.xlsx", row.names = FALSE)
write.xlsx(enriched[["Jensen_COMPARTMENTS"]], file = "DEG_Jensen_COMPARTMENTS_upreg_DAneuron6.12.xlsx", row.names = FALSE)



### Dot plots of genes of interest
# rename clusters
cluster_define <- as.numeric(as.character(merge.combined.sct@meta.data$seurat_clusters))
cluster_define[cluster_define == 1] <- "Xenograft DA"
cluster_define[cluster_define == 0] <- "Xenograft non-DA"
merge.combined.sct@meta.data$cluster_define <- cluster_define

cell_broad <- seurat_obj_subcells@meta.data$Cell_Type
cell_broad <- gsub("_.*","",cell_broad)
seurat_obj_subcells@meta.data$cell_broad <- cell_broad

# genes of interest
genes_to_plot <- unlist(strsplit(
  enriched$Jensen_COMPARTMENTS$Genes[
    which(enriched$Jensen_COMPARTMENTS$Term == "Synaptic membrane")],";"))

#expression data
exp_mat1 <- as.matrix(merge.combined.sct[["SCT"]]@counts)
row.names(exp_mat1) <- gsub("GRCh38--","",row.names(exp_mat1))
exp_mat1 <- exp_mat1[genes_to_plot,]

exp_mat2 <- as.matrix(seurat_obj_subcells[["SCT"]]@counts[genes_to_plot,])

# cell metadata
meta1 <- merge.combined.sct@meta.data

meta2 <- seurat_obj_subcells@meta.data
meta2 <- meta2[meta2$cell_broad %in% c("SOX6","CALB1","Inh","Ex"),]

# merge
meta1 <- bind_cols(meta1, as.data.frame(t(exp_mat1)))

exp_mat2 <- exp_mat2[,row.names(meta2)]
meta2 <- bind_cols(meta2, as.data.frame(t(exp_mat2)))

# select cols interest
meta1 <- meta1[,c("cluster_define",genes_to_plot)]
meta2 <- meta2[,c("cell_broad",genes_to_plot)]

# long format
meta1 <- pivot_longer(meta1, -cluster_define, names_to="Gene", values_to="Expression")
meta2 <- pivot_longer(meta2, -cell_broad, names_to="Gene", values_to="Expression")

# combine
colnames(meta2) <- colnames(meta1)
meta <- rbind(meta1,meta2)

# create summary
meta_summary <- meta %>%
  dplyr::group_by(cluster_define, Gene) %>%
  dplyr::summarise(Avg = mean(Expression),
            Pct = sum(Expression > 0) / length(Expression) * 100)

# order genes and cells
df_xeno <- meta_summary$Pct[meta_summary$cluster_define == "Xenograft DA"] - 
  meta_summary$Pct[meta_summary$cluster_define == "Xenograft non-DA"]
genes_to_plot_order <- meta_summary$Gene[meta_summary$cluster_define == "Xenograft DA"][order(df_xeno)]
meta_summary$Gene <- factor(meta_summary$Gene, levels=genes_to_plot_order)

meta_summary$cluster_define <- factor(meta_summary$cluster_define, levels=c("Inh","Ex","CALB1","SOX6",
                                                                            "Xenograft non-DA","Xenograft DA"))

# get meta_summaries
meta_summary$Avg <- log10(meta_summary$Avg)
dp1 <- dot_plot <- ggplot(meta_summary, aes(x=Gene, y=cluster_define)) +
  geom_point(aes(size = Pct, fill = Avg), color="black", shape=21) +
  scale_size("% detected", range = c(0,6)) +
  scale_fill_gradientn(colours = viridisLite::mako(100),
                       guide = guide_colorbar(ticks.colour = "black",
                                              frame.colour = "black"),
                       name = "Log10(Average\nexpression)") +
  ylab("") + xlab("") +
  theme(axis.text.x = element_text(size=10, angle=45, hjust=1, color="black"),
        axis.text.y = element_text(size=12, color="black"),
        axis.title = element_text(size=14)) +
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(axis.text.x=element_text(angle = 45, hjust = 1))

# save
arrange <- ggarrange(plotlist=list(dp1), nrow=2, ncol=2)
ggsave("dotplot_Synaptic.membrane.png", arrange)



## 2.  identify terms of interest
# terms interest
terms_interest <- toupper(c("Semaphorin","Netrin","Ret","WNT","Frizzled","L1CAM","Slit","Robo","Axon"))
tmp <- as.data.frame(enriched[["GO_Molecular_Function_2015"]])
matches <- unique(grep(paste(terms_interest,collapse="|"), 
                        toupper(tmp$Term)))

tmp[matches, ]

# axond guidance
axon <- read.xlsx("/Users/zacc/USyd/ASAP/snrna_seq/analysis/Axonal_guidance_Genes_List.xlsx")
tmp2 <- res[[1]]
upreg_deg <- tmp2[tmp2$logFC > 0,]

upreg_deg[row.names(upreg_deg) %in% axon$Name,]
axon[axon$Name %in%row.names(upreg_deg),]

genes_interest <- row.names(upreg_deg)[row.names(upreg_deg) %in% axon$Name]

enriched




# plot genes of interest
source("/Users/zacc/github_repo/ASAP-SpatialTranscriptomics/convenience/plotting_functions.R")

d1 <- merge.combined.sct@assays$SCT$counts
row.names(d1) <- gsub("GRCh38--","",row.names(d1))
m1 <- merge.combined.sct@meta.data

d2 <- seurat_obj_subcells@assays$SCT$counts
m2 <- seurat_obj_subcells@meta.data
d2 <- d2[,m2$Cell_Type %in% c("CALB1_CALCR","CALB1_CRYM_CCDC68","CALB1_GEM","CALB1_PPP1R17",            "CALB1_RBP4","CALB1_TRHR","SOX6_AGTR1","SOX6_DDT","SOX6_GFRA2","SOX6_PART1",
                              unique(m2$Cell_Type)[grep("Ex_|Inh_",unique(m2$Cell_Type))])]

m2 <- m2[colnames(d2),]
d1 <- as.data.frame(d1)
d2 <- as.data.frame(d2)
exp_dat <- merge(d1,d2,by="row.names")
row.names(exp_dat) <- exp_dat$Row.names
exp_dat <- exp_dat[,-1]

targ$cell_type <- targ$cell_type_1
targ$cell_type[targ$cell_type == 1] <- "Xenograft DA"
targ$cell_type[targ$cell_type == 0] <- "Xenograft non-DA"

gene_name <- "SEMA6D"
targ$gene <- unlist(exp_dat[gene_name,row.names(targ)])
cell_col <- viridis(length(unique(targ$cell_type_1)))

v1 <- violin_plot_nosig_function(targ,"cell_type","gene","Cell-Type",paste0(gene_name, " Norm Counts"), cell_col ) +
  theme(axis.text.x = element_text(angle = 45,  hjust=1))

gene_name <- "MAGI3"
targ$gene <- unlist(exp_dat[gene_name,row.names(targ)])
cell_col <- viridis(length(unique(targ$cell_type_1)))

v2 <- violin_plot_nosig_function(targ,"cell_type","gene","Cell-Type",paste0(gene_name, " Norm Counts"), cell_col ) +
  theme(axis.text.x = element_text(angle = 45,  hjust=1))

gene_name <- "ROBO2"
targ$gene <- unlist(exp_dat[gene_name,row.names(targ)])
cell_col <- viridis(length(unique(targ$cell_type_1)))

v3 <- violin_plot_nosig_function(targ,"cell_type","gene","Cell-Type",paste0(gene_name, " Norm Counts"), cell_col ) +
  theme(axis.text.x = element_text(angle = 45,  hjust=1))

arrange <- ggarrange(plotlist=list(v1,v2,v3), nrow=2, ncol=3, widths = c(2,2))
ggsave("gene_exp_examples.png", arrange)

# # downreg top 3 passing adj p-val < 0.05
# enriched <- enrichr(downreg, dbs)
# lapply(enriched, function(x) {
#   x <- x[x$Adjusted.P.value < 0.05,]
#   x[1:3,]
# })

# evaluate genes
tmp <- res[[3]]
tmp[c("SEMA6D","MAGI3","ROBO2"),]

tmp <- enriched[["GO_Molecular_Function_2015"]]

# evaluate axonal guidance members
tmp <- res[[1]]
axon <- read.xlsx("/Users/zacc/USyd/ASAP/snrna_seq/analysis/Axonal_guidance_Genes_List.xlsx")

tmp[row.names(tmp) %in% axon$Name,]



```










```